sled-agent/src/nexus.rs

```diff
@@ -6,3 +6,71 @@
 pub use crate::mocks::MockNexusClient as NexusClient;
 #[cfg(not(test))]
 pub use nexus_client::Client as NexusClient;
+
+use internal_dns_client::{
+    multiclient::{ResolveError, Resolver},
+    names::{ServiceName, SRV},
+};
+use omicron_common::address::NEXUS_INTERNAL_PORT;
+use slog::Logger;
+use std::net::Ipv6Addr;
+use std::sync::Arc;
+
+struct Inner {
+    log: Logger,
+    resolver: Resolver,
+}
+
+/// Wrapper around a [`NexusClient`] object, which allows deferring
+/// the DNS lookup until accessed.
+///
+/// Without the assistance of OS-level DNS lookups, the [`NexusClient`]
+/// interface requires knowledge of the target service IP address.
+/// For some services, like Nexus, this can be painful, as the IP address
+/// may not have even been allocated when the Sled Agent starts.
+///
+/// This structure allows clients to access the client on-demand, performing
+/// the DNS lookup only once it is actually needed.
+#[derive(Clone)]
+pub struct LazyNexusClient {
+    inner: Arc<Inner>,
+}
+
+impl LazyNexusClient {
+    pub fn new(log: Logger, addr: Ipv6Addr) -> Result<Self, ResolveError> {
+        Ok(Self {
+            inner: Arc::new(Inner {
+                log,
+                resolver: Resolver::new_from_ip(addr)?,
+            }),
+        })
+    }
+
+    pub async fn get(&self) -> Result<NexusClient, ResolveError> {
+        let address = self
+            .inner
+            .resolver
+            .lookup_ipv6(SRV::Service(ServiceName::Nexus))
+            .await?;
```

Comment:
Looking at [your code](https://github.com/oxidecomputer/omicron/pull/1347/files#diff-e58fdced2e69108a490916cca62b391c44fdde81d564de6f86b8d4e1febc5998R560-R561), we can probably add a method here to access the IP directly without generating the client - that would just be pulling this block of code into a `get_ip` method.